Integrating class symbol support

andyleejordan opened the pull request:
WIP: Attempting to integrate https://github.com/PowerShell/PowerShellEditorServices/pull/1886 with the changes previously made in https://github.com/PowerShell/PowerShellEditorServices/pull/1917.

fflaten committed:
add symbols for classes

fflaten committed:
remove code used for supporting PSv3 and v4

fflaten committed:
support hover

fflaten committed:
add type reference symboltype

fflaten committed:
support references and definitions

fflaten committed:
add codelens for class and enum

fflaten committed:
skip function ast visit for FunctionMemberAst

fflaten committed:
drop reference-support for ConfigurationDefinition

fflaten committed:
update document symbols test

fflaten committed:
refactor and cleanup

fflaten committed:
fix Contains extension method for same line extent

fflaten committed:
add EnumMember, signature and fix detection

fflaten committed:
update symbols test with name position

fflaten committed:
check type for property and function members

fflaten committed:
add symbolsservice tests

fflaten committed:
add codelens tests

fflaten committed:
add symbol tests using newline after keyword

fflaten committed:
fix DSC symbol test

andyleejordan committed:
Minor clean-ups

andyleejordan committed:
Remove deprecated (and confusing) `DocumentSymbolService.cs` file

andyleejordan committed:
Rename overloaded method to `ScanForReferencesOfSymbol`

Since `FindReferencesOfSymbol` is already used.

andyleejordan committed:
Migrate `FindReferencesVisitor` to `ReferenceTable.ReferenceVisitor`

This required using a `ScriptFile` for the associated test in
`AstOperationsTests`. The coupling of the visitor to the file seems fine
since that's the _point_ of this improvement: a cached sets of
references for each file. We also have to carefully sort our list of
symbol references for things that expect them in order.

andyleejordan committed:
Deduplicate `GetSymbolKind`, `GetDecoratedSymbolName`, and `GetRangeFromScriptRegion`

At least I'm pretty sure these were unnecessary duplicates.

andyleejordan commented:
@seeminglyscience with the integration pretty much all the changes you had to make to the unit tests I reverted. Can you explain if any of those were intentional (beyond just "making them pass")?

andyleejordan commented on src/PowerShellEditorServices/Services/Symbols/ReferenceTable.cs:
```diff
@@ -19,32 +21,46 @@ internal sealed class ReferenceTable
         return _symbolReferences.TryGetValue(command, out references);
     }
 
+    // TODO: Should this be improved, or pre-sorted?
```
@seeminglyscience a question for you.

andyleejordan commented on src/PowerShellEditorServices/Services/Symbols/ReferenceTable.cs:
```diff
@@ -55,64 +71,146 @@ internal void EnsureInitialized()
+        // TODO: We should examine if we really want to constrain the extents to the name only. This
+        // means that highlighting only highlights the symbol name, but providing the whole extend
+        // means the whole function (or class etc.) gets highlighted, which seems to be a personal
+        // preference.
```
Per https://github.com/PowerShell/PowerShellEditorServices/issues/1888 and https://github.com/PowerShell/PowerShellEditorServices/issues/1904 the current approach is valid (highlight only the name), so I need to remove this comment and we should add some tests.

andyleejordan commented on src/PowerShellEditorServices/Services/Symbols/ReferenceTable.cs:
```diff
@@ -55,64 +71,146 @@ internal void EnsureInitialized()
+            return AstVisitAction.Continue;
+        }
+
+        // TODO: What else can we implement?
```
Big open question here: how much else can we replace with this singular reference visitor class? Could we potentially identify when the "reference" is a definition, save that to the `SymbolReference`, and then use it later instead of searching all over again?

fflaten replied:
> Could we potentially identify when the "reference" is a definition, save that to the `SymbolReference`, and then use it later instead of searching all over again?

I suggested a `IsDefinition`-prop at one point here also. https://github.com/PowerShell/PowerShellEditorServices/issues/1889#issuecomment-1215560718

That might let us replace `FindDeclarationVisitor.cs`

andyleejordan replied:
This came along nicely!

andyleejordan commented on src/PowerShellEditorServices/Services/Symbols/SymbolsService.cs:
```diff
@@ -109,23 +110,18 @@ internal class SymbolsService
-                    reference.FilePath = scriptFile.FilePath;
-                    foundOccurrences.Add(reference);
-                }
+                // TODO: Each provider needs to set the source line and filepath.
```
So this irked me, a lot existing logic set the file path and source line at random times instead of during the initialization of the symbol reference. Which means I think something is broken.

andyleejordan commented on src/PowerShellEditorServices/Services/Symbols/SymbolsService.cs:
```diff
@@ -160,21 +156,20 @@ public List<SymbolReference> FindSymbolsInFile(ScriptFile scriptFile)
                 return null;
             }
 
+            // TODO: Should we handle aliases at a lower level?
```
Like for real we should do this better.

andyleejordan commented on src/PowerShellEditorServices/Services/Symbols/SymbolsService.cs:
```diff
@@ -227,36 +222,21 @@ static string[] GetIdentifiers(string symbolName, SymbolType symbolType, Command
                     cancellationToken.ThrowIfCancellationRequested();
                 }
+
+                symbolReferences.AddRange(fileReferences.OrderBy(symbol => symbol.ScriptRegion.StartOffset));
```
So sorting has been interesting. I don't know if VS Code cares, but the tests sure do.

andyleejordan commented on src/PowerShellEditorServices/Services/Symbols/SymbolsService.cs:
```diff
@@ -284,7 +264,12 @@ static string[] GetIdentifiers(string symbolName, SymbolType symbolType, Command
-            return AstOperations.FindReferencesOfSymbol(file.ScriptAst, foundSymbol).ToArray();
+            if (file.References.TryGetReferences(foundSymbol.SymbolName, out ConcurrentBag<SymbolReference> references))
+            {
+                return references.OrderBy(symbol => symbol.ScriptRegion.StartOffset).ToArray();
```
Ditto.

andyleejordan commented on test/PowerShellEditorServices.Test/Language/SymbolsServiceTests.cs:
```diff
@@ -154,9 +153,9 @@ public async Task FindsFunctionDefinitionForAlias()
-            Assert.Equal(2, referencesResult[0].ScriptRegion.StartColumnNumber);
+            Assert.Equal(3, referencesResult.Count);
+            Assert.Equal(1, referencesResult[0].ScriptRegion.StartLineNumber);
+            Assert.Equal(10, referencesResult[0].ScriptRegion.StartColumnNumber);
```
All these numbers essentially reverted to their values from before @seeminglyscience's cache work (they were changed in https://github.com/PowerShell/PowerShellEditorServices/pull/1917).

andyleejordan commented on test/PowerShellEditorServices.Test/Language/SymbolsServiceTests.cs:
```diff
@@ -285,7 +285,7 @@ public void FindsOccurrencesOnClass()
             IReadOnlyList<SymbolReference> occurrencesResult = GetOccurrences(FindsOccurrencesOnTypeSymbolsData.ClassSourceDetails);
             Assert.Equal(2, occurrencesResult.Count);
-            Assert.Equal("[SuperClass]", occurrencesResult[occurrencesResult.Count - 1].SymbolName);
+            Assert.Equal("SuperClass", occurrencesResult[occurrencesResult.Count - 1].SymbolName);
```
Except I don't know why the name for this changed.

fflaten replied:
Because you ordered `GetOccurrences()`. Don't see any code changed, so `TypeExpression` and `TypeConstraint` includes brackets while `TypeDefinition` (this) does not. :)

fflaten replied:
I have no idea. Maybe the tests didn't pass after all (can't remember). This seems correct though. 👍

andyleejordan commented on test/PowerShellEditorServices.Test/Language/SymbolsServiceTests.cs:
```diff
@@ -312,7 +312,7 @@ public void FindsOccurrencesOnEnum()
             IReadOnlyList<SymbolReference> occurrencesResult = GetOccurrences(FindsOccurrencesOnTypeSymbolsData.EnumSourceDetails);
             Assert.Equal(4, occurrencesResult.Count);
-            Assert.Equal("[MyEnum]", occurrencesResult[occurrencesResult.Count - 1].SymbolName);
+            Assert.Equal("MyEnum", occurrencesResult[occurrencesResult.Count - 1].SymbolName);
```
Ditto.

fflaten replied:
Same as above

andyleejordan commented on test/PowerShellEditorServices.Test.Shared/References/FunctionReference.ps1:
```diff
@@ -0,0 +1,24 @@
+function BasicFunction {}
```
I had to make this a file so that the test can get a `ScriptFile`... which has the `References` field which runs the new cached `ReferenceVisitor`.

andyleejordan reviewed (commented):
My own questions for you both.

andyleejordan commented:
Hm ok so the sorting isn't quite right. Debugging `FindsReferencesOnTypeConstraint` and while all four references are found, they're out of order because the one used as the param type constraint has a `StartOffset` of 0.

andyleejordan committed:
Move reference sorting to unit tests

Since the client doesn't actually care which order we return the
references in, we should skip unnecessarily sorting. It appears to apply
its own sorting to the results regardless of their order, and the
specification doesn't say it needs to be sorted.

Also adjust two tests to account for the definitions being in the
reference lists.

andyleejordan committed:
More simplification and dead code elimination

andyleejordan committed:
Track declarations on `SymbolType` so we can use `ReferenceTable` more

This lets us (nearly) replace `FindSymbolsInDocument` by using the references
(symbols) from the `ScriptFile`.

andyleejordan committed:
Integrate `FindSymbolsVisitor` into `ReferenceTable`

We were missing the distinction between properties and enum members,
as well as DSC configuration.

andyleejordan committed:
Actually fix sorting in symbols tests

`StartOffset` is always 0, sort by `Range.Start` instead.

andyleejordan committed:
Make `DocumentSymbolHandler` more readable

While we're debugging it.

andyleejordan committed:
Fix outline view (filter out `EmptyScriptExtent`)

And also prepare `SymbolReference` for eventual support of `DocumentSymbol`.

fflaten commented:
Currently shows variable-assignments for every parameter and re-assignments inside functions/methods in outline + workspace search. Feels messy compared to stable. 

Maybe stick with current behavior for now? If so, it might require either of these:
- Sticking with `FindSymbolsVisitor` for documentsymbols and workspacesymbols handlers.
- Add another flag or bool in `SymbolReference` to control visibility in the handlers
- Include a `Parent`-backlink in the `SymbolReference` we can filter on in handler (to keep only root-level). This is easier to achieve now with a single reference visitor and is probably required to support `DocumentSymbol.children` anyway, so I'd prefer this.

fflaten commented on src/PowerShellEditorServices/Services/Symbols/SymbolsService.cs:
```diff
@@ -284,7 +263,12 @@ static string[] GetIdentifiers(string symbolName, SymbolType symbolType, Command
             }
 
-            return AstOperations.FindReferencesOfSymbol(file.ScriptAst, foundSymbol).ToArray();
+            if (file.References.TryGetReferences(foundSymbol.SymbolName, out ConcurrentBag<SymbolReference> references))
```
Filter on symbol type. Will currently match ex. a similar function name and enum/class name

fflaten commented on test/PowerShellEditorServices.Test/Language/SymbolsServiceTests.cs:
```diff
@@ -256,10 +255,11 @@ public void FindsOccurrencesOnParameter()
-            Assert.Equal("Get-ChildItem", referencesResult[referencesResult.Count - 1].SymbolName);
+            Assert.Collection(referencesResult,
+                (i) => Assert.Equal("Get-ChildItem", i.SymbolName),
+                (i) => Assert.Equal("gci", i.SymbolName),
```
👍 Why was this changed in https://github.com/PowerShell/PowerShellEditorServices/pull/1917 in the first place? Never made sense to me, but didn't investigate.

andyleejordan replied:
Ditto to asking why and not being sure, but not investigating either.

fflaten commented on src/PowerShellEditorServices/Services/Symbols/SymbolsService.cs:
```diff
@@ -284,7 +263,12 @@ static string[] GetIdentifiers(string symbolName, SymbolType symbolType, Command
             }
 
-            return AstOperations.FindReferencesOfSymbol(file.ScriptAst, foundSymbol).ToArray();
+            if (file.References.TryGetReferences(foundSymbol.SymbolName, out ConcurrentBag<SymbolReference> references))
+            {
+                return references;
+            }
```
```suggestion
            if (file.References.TryGetReferences(foundSymbol.SymbolName, out ConcurrentBag<SymbolReference> references))
            {
return references.Where((r) =>
                    foundSymbol.SymbolType == r.SymbolType ||
                    (foundSymbol.SymbolType is SymbolType.Class or SymbolType.Enum or SymbolType.Type
                        && r.SymbolType is SymbolType.Class or SymbolType.Enum or SymbolType.Type)
                );
            }
```

Filter on symbol type. Will currently match ex. a function and enum/class/type with same name.

andyleejordan commented:
> * Include a `Parent`-backlink in the `SymbolReference` we can filter on in handler (to keep only root-level). This is easier to achieve now with a single reference visitor and is probably required to support `DocumentSymbol.children` anyway, so I'd prefer this.

Same, that's what I was thinking.

andyleejordan committed:
Fix typo

andyleejordan committed:
Replace `FindDeclarationVisitor`

fflaten commented on src/PowerShellEditorServices/Services/Symbols/ReferenceTable.cs:
```diff
@@ -55,62 +71,228 @@ internal void EnsureInitialized()
+            return AstVisitAction.Continue;
+        }
+
+        public override AstVisitAction VisitInvokeMemberExpression(InvokeMemberExpressionAst methodCallAst)
```
Which scenario is this for? I don't think you'll be able to match it to the correct method to count it as a reference.

andyleejordan replied:
So it turns out it works quite well! By bucketing all the overloads (of the same name) together, and returning a set of possible definitions, VS Code just does what it needs so you can do:

```powershell
class SuperClass : BaseClass {
    [string]MyClassMethod([string]$param1, $param2, [int]$param3) {
        $this.SomePropWithDefault = 'something happend'
        $param1
        return 'finished'
    }

    [string]
    MyClassMethod([MyEnum]$param1) {
        return 'hello world'
    }
    [string]MyClassMethod() {
        return 'hello world'
    }
}

$o = [SuperClass]::new()
$o.MyClassMethod()
```

and find definition on `MyClassMethod()` shows the three potential matches:
<img width="1681" alt="Screenshot 2023-01-20 at 5 48 12 PM" src="https://user-images.githubusercontent.com/2226434/213835617-cac4d752-3b3e-4e46-8966-a8c7c79985fb.png">

No, it's not perfect, but it's a _whole lot better_ than nothing!

fflaten replied:
True, go to definition makes sense. 👍 You could also filter on argument count, but argument types won't work.

fflaten replied:
To make it complete, should `::new()` return the constructors? Because that's a reserved method when static (`InvokeMemberExpressionAst.Static == true`), right?

andyleejordan replied:
That's a good idea!

andyleejordan committed:
Remove notion of `ReferencedFiles`

This has been unused and sometimes broken,
it doesn't actually power any useful features.

andyleejordan committed:
Break the soon-to-be-only visitor out

andyleejordan committed:
Remove `FindCommandVisitor`, `FindSymbolVisitor` and `FindDotSourcedVisitor`

`SymbolsService` is now nullable too.

andyleejordan committed:
Replace `VisitCommandParameter` with `VisitParameter`

andyleejordan committed:
Mark Pester symbols as declarations

dependabot[bot] committed:
Bump Microsoft.PowerShell.SDK from 7.3.1 to 7.3.2 (#1986)

Bumps [Microsoft.PowerShell.SDK](https://github.com/PowerShell/PowerShell) from 7.3.1 to 7.3.2.
- [Release notes](https://github.com/PowerShell/PowerShell/releases)
- [Commits](https://github.com/PowerShell/PowerShell/compare/v7.3.1...v7.3.2)

---
updated-dependencies:
- dependency-name: Microsoft.PowerShell.SDK
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

andyleejordan committed:
Clean up `SymbolReference` constructors

andyleejordan committed:
Fix file names for visitors

andyleejordan committed:
Clean up some `IEnumerables`

It's so sad we can't use `IAsyncEnumerable`.

andyleejordan committed:
Add `DisplayName` to clean up outline

andyleejordan committed:
Exclude parameters from workplace symbols

andyleejordan committed:
Optionally open all workspace files and scan for symbols

andyleejordan committed:
Use `NameRegion` for definition, workplace jump and hover

andyleejordan committed:
Fix jump to definition from outline

By starting at the start of `NameRegion`.

Co-authored-by: Frode Flaten <3436158+fflaten@users.noreply.github.com>

andyleejordan committed:
Improve function's display strings

andyleejordan committed:
Filter symbol references to type matches

andyleejordan committed:
Short-circuit find-definition when given definition

andyleejordan committed:
Decorate more type names

And guard from null exception in definition handler.

andyleejordan committed:
Include parameter type constraints as symbols

andyleejordan committed:
Ignore some warnings

andyleejordan committed:
Fix parameter hints tests (but the whole feature needs work)

andyleejordan committed:
Fix find function definition tests

andyleejordan committed:
Fix find variable and function references tests

andyleejordan committed:
Fix `FindsOccurrencesOnParameter` test

andyleejordan committed:
Fix `FindsReferencesOnCommandWithAlias` test

andyleejordan committed:
Fix a bunch of class symbol tests

andyleejordan committed:
Fix `FindsPropertyDefinition` test

Name was wrong.

andyleejordan committed:
Fix property symbols tests

andyleejordan committed:
Fix `ScriptRegion.ContainsPosition`

To actually be an inclusive check of the given point being
between the two boundary points.

andyleejordan committed:
Fix display string for function definitions

Space between name and subsequent parenthesis.

andyleejordan committed:
Fix find enum member tests

andyleejordan committed:
Add symbol type equivalencies test

andyleejordan committed:
Remove deprecated dot-source reference tests (and data)

Let it be known that Patrick bets $500 that no one notices this.
And to be fair, searching the whole workspace now works great.

andyleejordan committed:
Fix `FindsDetailsForBuiltInCommand` test

And delete "find details" for non-commands (as they all get their
display strings set on creation now).

andyleejordan committed:
Fix the rest of the symbols service tests

andyleejordan committed:
Run `FindsDetailsForBuiltInCommand` only on macOS

andyleejordan committed:
Skip `FindsSymbolsInDSCFile` as DSC symbols don't yet work

The visitor exists but it's not called, probably a module-import issue.

andyleejordan committed:
Mark some methods internal

Co-authored-by: Patrick Meinecke <SeeminglyScience@users.noreply.github.com>

andyleejordan committed:
Rename `SymbolReference` fields

andyleejordan committed:
Replace `SymbolTypeMatches` logic with prefixed identifiers

So that we rely on the dictionary to sort our symbols into the equivalent
types instead of having to perform a filter.

andyleejordan committed:
Strip qualifier from variable name in symbol identifier

andyleejordan committed:
Handle symbol prefix identifier in `CommandHelpers`

I think there's a cleaner way to do this, but this works.

andyleejordan committed:
More leniently filter workspace symbols

Against the full name and not just the identifier, since it's
what's displayed in the search menu.

Now you can search methods by their parameters' names.

andyleejordan committed:
Map symbol types to LSP better

andyleejordan committed:
Restore `public ScriptRegion Create(...)` method

andyleejordan committed:
Fix bugs in `VisitorUtils` around `LastIndexOf`

Should have been just `IndexOf`, and also use `+` instead
of `string.Concat()` for simplicity and speed.

andyleejordan committed:
Distinguish properties/enums from variables

andyleejordan committed:
Distinguish constructors/methods from functions

andyleejordan committed:
Remove unused code from `Extensions.cs`
